File: python/arrays-and-hashing/first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            correct = nums[i] - 1
            while nums[i] > 0 and nums[i] < len(nums) and nums[correct] != nums[i]:
                nums[i], nums[correct] = nums[correct], nums[i]
                correct = nums[i] - 1

        for i in range(len(nums)):
            if i + 1 != nums[i]:
                return i + 1

        return len(nums) + 1

File: c++/arrays-and-hashing/first-missing-positive.cc
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for (int i = 0; i < nums.size(); ++i) {
            while (nums[i] < nums.size() && nums[i] > 0 && nums[i] != nums[nums[i] - 1]) {
                std::swap(nums[nums[i] - 1], nums[i]);
            }
        }

        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        
        return nums.size() + 1;
    }
};
